stan/math/rev/functor/ode_adjoint.hpp

```diff
@@ -0,0 +1,168 @@
+ * through to \p f without modification).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_y0 Type of initial state
+ * @tparam T_t0 Type of initial time
+ * @tparam T_ts Type of output times
+ * @tparam T_Args Types of pass-through parameters
+ *
+ * @param function_name Calling function name (for printing debugging messages)
+ * @param f Right hand side of the ODE
+ * @param y0 Initial state
+ * @param t0 Initial time
+ * @param ts Times at which to solve the ODE at. All values must be sorted and
+ *   not less than t0.
+ * @param relative_tolerance_forward Relative tolerance for forward problem
+ * passed to CVODES
+ * @param absolute_tolerance_forward Absolute tolerance per ODE state for
+ * forward problem passed to CVODES
+ * @param relative_tolerance_backward Relative tolerance for backward problem
+ * passed to CVODES
+ * @param absolute_tolerance_backward Absolute tolerance per ODE state for
+ * backward problem passed to CVODES
+ * @param relative_tolerance_quadrature Relative tolerance for quadrature
+ * problem passed to CVODES
+ * @param absolute_tolerance_quadrature Absolute tolerance for quadrature
+ * problem passed to CVODES
+ * @param max_num_steps Upper limit on the number of integration steps to
+ *   take between each output (error if exceeded)
+ * @param num_steps_between_checkpoints Number of integrator steps after which a
+ * checkpoint is stored for the backward pass
+ * @param interpolation_polynomial type of polynomial used for interpolation
+ * @param solver_forward solver used for forward pass
+ * @param solver_backward solver used for backward pass
+ * @param[in, out] msgs the print stream for warning messages
+ * @param args Extra arguments passed unmodified through to ODE right hand side
+ * @return Solution to ODE at times \p ts
+ */
+template <typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args, require_eigen_col_vector_t<T_y0>* = nullptr>
+std::vector<Eigen::Matrix<stan::return_type_t<T_y0, T_t0, T_ts, T_Args...>,
+                          Eigen::Dynamic, 1>>
+ode_adjoint_impl(const char* function_name, const F& f, const T_y0& y0,
+                 const T_t0& t0, const std::vector<T_ts>& ts,
+                 double relative_tolerance_forward,
+                 const Eigen::VectorXd& absolute_tolerance_forward,
+                 double relative_tolerance_backward,
+                 const Eigen::VectorXd& absolute_tolerance_backward,
+                 double relative_tolerance_quadrature,
+                 double absolute_tolerance_quadrature,
+                 long int max_num_steps,                  // NOLINT(runtime/int)
+                 long int num_steps_between_checkpoints,  // NOLINT(runtime/int)
+                 int interpolation_polynomial, int solver_forward,
+                 int solver_backward, std::ostream* msgs,
+                 const T_Args&... args) {
+  auto integrator = new stan::math::cvodes_integrator_adjoint_vari<
```

Comment:
A vari shouldn't get created if there's no autodiff involved because noone would expect varis to get created.

The arena types will also allocate on the stack which shouldn't happen for the all arithmetic case.

What are our options here?

Reply:
I would leave it as is.... since the stiff solver (also Adams which is implicit) will anyway trigger nested AD.

In case it's a requirement for a double return type to not put vari's on the stack we could possibly implement a logic where we create the vari on a nested stack and then return the double solution?

Reply:
It might be easier to just copy-paste a version of this and just start deleting autodiff stuff until all that is left is forward mode.

Kinda annoying to have duplicate code, but if the ode test system does the thing where it tests the arithmetic version fully and then tests the autodiff version against the arithmetic version, we're good to go.

A nested autodiff solution wouldn't be the worst thing. Seems a bit weird but it handles the problem of accidentally allocating a bunch of stuff in prim-only execution. I'd prefer the duplicate-code of those two solutions.

Reply:
Why couldn't we just take the stuff from inside the constructor, make it all double types, and then make a new function to return that? The values for the return even in the vari case should never be changed after the forward pass right?

Reply:
A possible issue with this is you do special stuff in the forward pass if you're gonna need the backward pass, so making everything totally generic might be a little tricky (like check the if [here](https://github.com/stan-dev/math/pull/1905/files#diff-e41bca45735d92ce6c146715d8d3c09ba22c2c1beb279cf4776becd41d0e2e08R327))